### DISCUSSION TITLE
Define a schema which validates a class instance has a method shaped a specific way?

### DISCUSSION THREAD

**Discussion by goleary** [colinhacks/zod#531] [Q&A] [answered]
We use and love `zod` at https://www.getupnext.com/. One of the things we use it for is validating data that is coming out of [Cloud Firestore](https://firebase.google.com/docs/firestore), a schemaless NoSQL db. For whatever reason, Firestore does not support dates, so any dates that you send to Firestore are transformed into their own `Timestamp` class. I want a schema that supports validating the data coming out of firestore (where properties will be `instance of Timestamp`).

Here's what I came up with:
```ts
const TimestampToDate = z.instanceof(firebase.firestore.Timestamp).transform((t) => t.toDate());

const Thingy = z.object({
  createdAt: TimestampToDate.or(z.date()),
});
```

Here's the kicker though, we use firestore in multiple environments (web, server, mobile) & the `Timestamp` class is different in the different environments, so this schema can not be used across environments to validate the same data because the instance is different in each environment.

Even though the class differs across environments, in all environments is `Timestamp` is a class which has a function called `toDate` which returns a `Date`. Because of this I tried another option which I was hoping would work:

```ts
import { z } from 'zod';
import firebase from 'firebase-admin';

const Timestamp = z.object({
  toDate: z.function().returns(z.date()),
});
const TimestampToDate = Timestamp.transform((t) => t.toDate());

const Thingy = z.object({
  createdAt: TimestampToDate,
});

const thingyFromFirestore = Thingy.parse({
  createdAt: new firebase.firestore.Timestamp(1000, 0),
});
```

But this does not work because `zod` treats `createdAt` as an object instead of a class instance (totally understandable), and the reference to `this` is lost:

```
  return new Date(this._seconds * 1000 + Math.round(this._nanoseconds / MS_TO_NANOS));
TypeError: Cannot read property '_seconds' of undefined
```

@colinhacks Is it currently possible to define a schema which validates that a class instance has a method shaped a specific way, but ignore the entire rest of the instance?

If not is this something you have thought about supporting?

Something along the lines of:

```ts
const Timestamp = z.instanceofWithProp({
  createdAt: TimestampToDate,
});
```

Thanks for reading & creating this awesome library!

**Comment by colinhacks** [comment, marked as answer]
I'd recommend something like this

```ts
const Timestamp = z.custom<{toDate:()=>Date}>();
const TimestampToDate = Timestamp.transform((t) => t.toDate()).or(z.date());
```

Under the hood, `z.custom` delegates to `z.any`, which is noteworthy in that it returns the exact object that's passed in (as opposed to z.object, which returns a deep clone, as you alluded to). Currently `Timestamp` allows any value, though you could provide some simple validation like so:

```ts
const Timestamp = z.custom<{toDate:()=>Date}>((val)=>{
  return val && typeof val['toDate'] === "function"
});
```
It's not possible to validate the return type of `toDate` so this just checks that a `toDate` method exists.

**Reply by goleary** [reply]
That worked great, thanks for the suggestions!

One question:

Is it not possible to validate that what is returned by `toDate()` is actually a date by calling the method?

```ts
const Timestamp = z.custom<{toDate:()=>Date}>((val)=>{
  return val && typeof val['toDate'] === "function" && val.toDate() instanceof Date;
});
```

This seems to me to work as intended, throwing an error when `toDate()` returns something other than a date, but perhaps I'm missing something?

**Reply by goleary** [reply]
Or we could "chain" the transform output into a date schema parse?

```ts
const TimestampToDate = Timestamp.transform((t) =>
  z.date().parse(t.toDate())
).or(z.date());
```
this also seems to work, but perhaps there is a more elegant way to represent it?

**Reply by colinhacks** [reply]
That's as elegant as it gets. There's no way to define a post-transform "endpoint" schema currently.

**Comment by davidoort** [comment]
Hey guys, I'm having issues with Zod + Firestore timestamps too.
Here's where I'm getting stuck:

```typescript
type ModelExample = {
  date: FirebaseFirestore.Timestamp;
};

const ModelExampleSchema: toZod<ModelExample> = z.object({
  // None of these work
  // date: z.instanceof(FirebaseFirestore.Timestamp),
  // date: z.date(),
  // date: TimestampToDate,
});
```

Would appreciate any ideas @colinhacks @goleary 🙏

Looking forward to getting this to work so I can use the full power of Zod!
